### DISCUSSION TITLE
Are *.pyi files preferred to py.typed?

### DISCUSSION THREAD

**Discussion by kyle-query** [microsoft/pyright#4963] [Q&A] [answered]
My team maintains a large library that doesn't include any *.pyi files, however the *.py files are fully annotated. Is there any benefit to generating *.pyi files, compared to packaging the py.typed file in the module root directory?

**Comment by erictraut** [comment, marked as answer]
Thanks for posting the question. For detailed guidance about providing type information in libraries, please refer to [this documentation](https://microsoft.github.io/pyright/#/typed-libraries).

In short, we recommend inlined types (i.e. types included in ".py" files) rather than separate ".pyi" files. Inlined types are easier to maintain, and ".py" files typically contain more information (including doc strings and default argument values) which is useful when the library is used with a language server.

The only benefit of separate ".pyi" files is that they are smaller than the corresponding ".py" files, so a static type evaluator might be able to more quickly load and parse the files, but this performance improvement is minimal, typically not perceptible.
